homeassistant/components/device_tracker/o2box.py

```diff
@@ -0,0 +1,226 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_scanner(hass, config):
+    """ Validates config and returns O2BoxScanner. """
+    if not validate_config(config,
+                           {DOMAIN: []},
+                           _LOGGER):
+        return None
+
+    scanner = O2BoxScanner(config[DOMAIN])
+    return scanner if scanner.success_init else None
+
+
+class O2BoxScanner(object):
+    """This class queries the O2Box"""
+    def __init__(self, config):
+        self.last_results = []
+        self.password = ''
+        self.success_init = True
+
+        # Check for user specific configuration
+        if CONF_HOST in config.keys():
+            self.host = config[CONF_HOST]
+        if CONF_PASSWORD in config.keys():
+            self.password = config[CONF_PASSWORD]
+
+        # Establish a connection to the FRITZ!Box
+        self.o2_box = O2Box(host=self.host, routerpassword=self.password)
+
+        if self.o2_box.try_login():
+            self.success_init = True
+            _LOGGER.info("Successfully connected to O2-Box")
+            self._update_info()
+        else:
+            self.success_init = False
+            _LOGGER.error("Failed to establish connection to O2Box "
+                          "with IP: %s", self.host)
+
+    @Throttle(MIN_TIME_BETWEEN_SCANS)
+    def _update_info(self):
+        """ Retrieves latest information from the FRITZ!Box. """
+        if not self.success_init:
+            return False
+
+        _LOGGER.info("Scanning")
+        self.last_results = self.o2_box.get_wireless_devices()
+        return True
+
+    def scan_devices(self):
+        """ Scan for new devices and return a list of found device ids. """
+        self._update_info()
+        active_hosts = []
+        for known_host in self.last_results:
+            active_hosts.append(known_host.mac)
+        return active_hosts
+
+    def get_device_name(self, mac):
+        """ Returns the name of the given device or None if is not known. """
+        self._update_info()
+        for known_host in self.last_results:
+            if known_host.mac == mac:
+                return known_host.name
+
+        return None
+
+
+class O2Box(object):
+    """This is the class which connects over HTTP to the O2Box
+    and parses the HTML to extract information"""
+    def __init__(self, host, routerpassword):
+        self.baseurl = "http://{}".format(host)
+        self.password = routerpassword
+
+    @staticmethod
+    def _pretty_mac(macugly):
+        macs = macugly.replace('[', '').replace(']', '').replace('\'', '').split(',')
+        return ':'.join(macs)
+
+    def _extract_wireless_info(self, lines):
+        """
+        Extracts the info of all the wireless devices from the Javascript section of the HTML
+        """
+        clientlines = list(filter(lambda l: 'STA_infos[' in l and 'lan_client_t' not in l, lines))
+        client_cnt = int(len(clientlines) / 4)
+
+        def extract_client_infos():
```

Comment:
Why is this broken out into a sub-function? It seems as if this abstraction is unnecessary.